ttnn/cpp/ttnn/device_operation.hpp

```diff
@@ -0,0 +1,311 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <concepts>
+#include <optional>
+#include <tt_eager/tensor/tensor.hpp>
+
+#include "third_party/magic_enum/magic_enum.hpp"
+#include "tt_dnn/op_library/operation_history.hpp"
+#include "tt_metal/third_party/tracy/public/tracy/Tracy.hpp"
+#include "tt_metal/tools/profiler/op_profiler.hpp"
+#include "tt_stl/concepts.hpp"
+#include "tt_stl/reflection.hpp"
+#include "tt_stl/unique_any.hpp"
+
+namespace ttnn {
+
+namespace device_operation {
+
+template <typename... program_attributes_t>
+struct CachedProgram {
+    tt::tt_metal::Program program;
+    // Cached program needs to share program_attributes between create and override_runtime_arguments functions
+    std::tuple<program_attributes_t...> program_attributes;
+
+    CachedProgram(tt::tt_metal::Program&& program, program_attributes_t... program_attributes) :
+        program{std::move(program)}, program_attributes{std::tuple{program_attributes...}} {}
+};
+
+struct CachedProgramFactory {
+    tt::stl::unique_any<1024, 32> cached_program;
+    std::size_t program_factory_index;
+
+    template <typename... program_attributes_t>
+    CachedProgramFactory(CachedProgram<program_attributes_t...>&& cached_program, std::size_t program_factory_index) :
+        cached_program{std::move(cached_program)}, program_factory_index{program_factory_index} {}
+};
+
+template <typename program_factory_t>
+concept ProgramFactoryConcept = requires {
+    [](const auto& operation_attributes, const auto& tensor_args, auto& tensor_return_value) {
+        auto cached_program = program_factory_t::create(operation_attributes, tensor_args, tensor_return_value);
+        program_factory_t::override_runtime_arguments(
+            cached_program, operation_attributes, tensor_args, tensor_return_value);
+    };
+};
+
+template <typename operation_t>
+concept DeviceOperationConcept = requires {
+    [](const typename operation_t::operation_attributes_t& operation_attributes,
+       const typename operation_t::tensor_args_t& tensor_args) {
+        operation_t::validate_on_program_cache_hit(operation_attributes, tensor_args);
```

Comment:
we now require `DeviceOperation` to implement both `operation_t::validate_on_program_cache_hit` and `operation_t::validate_on_program_cache_miss`? Should we add some logging static asserts?

Reply:
sounds good